src/psyclone/psyGen.py

```diff
@@ -1357,11 +1379,342 @@ def dag_name(self):
+        if len(nodes) != 1:
+            raise GenerationError(
+                "A Schedule containing an ACC parallel region must also "
+                "contain an ACC enter data directive but none was found for "
+                "{0}".format(self.root.invoke.name))
+        # Check that the enter-data directive comes before this parallel
+        # directive
+        if nodes[0].abs_position > self.abs_position:
+            raise GenerationError(
+                "An ACC parallel region must be preceeded by an ACC enter-"
+                "data directive but in {0} this is not the case.".
+                format(self.root.invoke.name))
+
+        # "default(present)" means that the compiler is to assume that
+        # all data required by the parallel region is already present
+        # on the device. If we've made a mistake and it isn't present
+        # then we'll get a run-time error.
+        parent.add(DirectiveGen(parent, "acc", "begin", "parallel",
+                                "default(present)"))
+
+        for child in self.children:
+            child.gen_code(parent)
+
+        parent.add(DirectiveGen(parent, "acc", "end", "parallel", ""))
+
+    @property
+    def ref_list(self):
+        '''
+        Returns a list of the references (whether to arrays or objects)
+        required by the Kernel call(s) that are children of this
+        directive. This is the list of quantities that must be
+        available on the remote device (probably a GPU) before
+        the parallel region can be begun.
+
+        :returns: list of variable names
+        :rtype: list of str
+        '''
+        variables = []
+
+        # Look-up the calls that are children of this node
+        for call in self.calls():
+            for arg in call.arguments.acc_args:
+                if arg not in variables:
+                    variables.append(arg)
+        return variables
+
+    @property
+    def fields(self):
+        '''
+        Returns a list of field objects required by the Kernel call(s) that are
+        children of this directive.
+
+        :returns: list of names of field arguments
+        :rtype: list of str
+        '''
+        # Look-up the calls that are children of this node
+        obj_list = []
+        for call in self.calls():
+            for arg in call.arguments.fields:
+                if arg not in obj_list:
+                    obj_list.append(arg)
+        return obj_list
+
+    @property
+    def scalars(self):
+        '''
+        Returns a list of the scalar quantities required by the Calls in
+        this region.
+
+        :returns: list of names of scalar arguments
+        :rtype: list of str
+        '''
+        scalars = []
+        for call in self.calls():
+            for arg in call.arguments.scalars:
+                if arg not in scalars:
+                    scalars.append(arg)
+        return scalars
+
+
+class ACCLoopDirective(ACCDirective):
```

Comment:
Apparently it's assumed if the loop directive is within a parallel region. However, I've added support for it as it might be useful in future (e.g. within kernels).